Update RenderNode and RenderIndex specs for versioning/diffing

mayaepps opened the pull request:
Bug/issue #, if applicable: 

## Summary

This updates the RenderNode spec for archive versioning using JSON Patches. It also updates the RenderIndex spec for versioning using JSON Patches, as diffing using a dictionary that maps from all previous versions of a given Node to what type of change should be displayed (such as “modified,” or “added”) when diffing between the current version and each past version.

## Testing

Copy and paste each spec into an OpenAPI spec validator.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [n/a?] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [n/a] Updated documentation if necessary

Commit:
Add version diffing properties to RenderNode spec

Adds a `versions` property to the RenderNode, which is an array of VersionPatches. Each VersionPatch has an archive version that the contained JSONPatch converts the RenderNode back to.
This commit also adds versionID and version display name to the metadata, and adds display name to Info.plist.

ethan-kusters commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderNode.spec.json:
```diff
@@ -1501,11 +1508,11 @@
                     "kind": {
                         "type": "string",
                         "enum": [
-                            "article", 
-                            "project", 
-                            "section", 
-                            "overview", 
-                            "symbol", 
+                            "article",
+                            "project",
+                            "section",
+                            "overview",
+                            "symbol",
```
I think it would be better to leave this as-is just to not confuse the commit history.

ethan-kusters commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderNode.spec.json:
```diff
@@ -2856,6 +2870,18 @@
+                "required": ["version", "patch"],
+                "properties": {
+                    "version": {
+                        "$ref": "#/components/schemas/SchemaVersion"
```
Should we require the version to be a specific format or just require this to be a unique string? 

I'm also wondering if we should have a separate display name?

For example, maybe I might want to use a commit hash as my unique version identifier, but then provide "Xenial" ([an ubuntu version name](https://en.wikipedia.org/wiki/Ubuntu_version_history#Ubuntu_16.04_LTS_(Xenial_Xerus))) as the display name.

Not entirely opposed to enforcing Semantic Versioning here though- just curious what your thoughts are around this.

mayaepps replied:
I like that idea. I thought using Semantic Versioning would be pretty easy since it already exists in every RenderNode, so it would be easy to grab from there (and I was imagining it would be unique), but it seems more flexible to have a unique version identifier and a display name. And instead of having the user select two semantic versions to diff between, it sounds more readable to select two display names like Xenial and Yakkety. Would the user have to pass all the version IDs and display names in on the command line then?

ethan-kusters replied:
> Would the user have to pass all the version IDs and display names in on the command line then?

That's a good question. I think they we should allow for passing them via the command-line for automation use-cases but we already have a version property in the `Info.plist` file that can be included in a DocC catalog so I think we could use that as well: https://github.com/apple/swift-docc/blob/main/Sources/DocCDocumentation/DocCDocumentation.docc/Info.plist#L24.

And maybe add an additional `VersionDisplayName` property to that spec.

ethan-kusters reviewed (commented):
This looks great! Awesome work. Just left a couple questions.

ethan-kusters commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderNode.spec.json:
```diff
@@ -2856,6 +2870,21 @@
+                "type": "object",
+                "required": ["versionID", "displayName", "patch"],
+                "properties": {
+                    "versionID": {
```
Something that occurred to me- I think we'll need to specify a version identifier and display name for the base RenderNode as well.

Maybe we should make a new `Version` object that includes `identifier` and `displayName` properties? And then include that in the RenderNode's `metadata` as well as in each `VersionPatch`. What do you think?

mayaepps replied:
I'm going back and making these changes and noticed you resolved this comment--just making sure we're still doing this?

ethan-kusters commented on Sources/DocCDocumentation/DocCDocumentation.docc/Info.plist:
```diff
@@ -22,6 +22,8 @@
 	<string>0.1.0</string>
 	<key>CFBundleVersion</key>
 	<string>0.1.0</string>
+	<key>VersionDisplayName</key>
+	<string>0.1.0</string>
```
I think we can hold off on making this change until we actually adopt versioning in DocC's documentation. But definitely something we should look at doing as part of this project in a future PR.

ethan-kusters reviewed (approved):
Looks great!! 🚀

ethan-kusters commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderIndex.spec.json:
```diff
@@ -99,6 +105,16 @@
                         }
+                    },
+                    "diffs": {
+                        "type": "object",
```
Maybe we should add some documentation to this type since it's a little tricky to understand? (You can use the `summary` and `description` OpenAPI properties: https://swagger.io/docs/specification/basic-structure/)

I think we should also specify that if there's nothing changed against a given version, that version just won't appear in the `diffs` dictionary.

ethan-kusters commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderIndex.spec.json:
```diff
@@ -99,6 +105,16 @@
                             "$ref": "#/components/schemas/Node"
                         }
+                    },
+                    "diffs": {
```
I'm struggling to come up with a better alternative but `diffs` seems a little too generic to me... What do you think?

Maybe `versionChanges` or `versionDifferences`?

mayaepps replied:
I agree and I’m also struggling with a better alternative (`previousNodeDifferences`? 🤨)--I like `versionDifferences`, I just want to be careful that something like `versionDifferences` isn’t easily confused with the `versions` that contains patches (maybe this means the `versions` property name should be changed to be more specific?) Or maybe it won’t be too confusing since they will show up in different contexts?

mayaepps commented on Sources/SwiftDocC/SwiftDocC.docc/Resources/RenderIndex.spec.json:
```diff
@@ -26,6 +27,31 @@
+                        }
+                    },
+                    "nodeVersionDifferences": {
+                        "description": "A dictionary property that maps from a node identifier to a dictionary of diffs for that node. The node diffs dictionary maps from a version to the type of change (modified/added) the current node should display when diffing against that prior version. If nothing changed against a given version, that version won't appear in the dictionary of diffs.",
```
For some reason, adding a `summary` property as well causes [the validator](https://apitools.dev/swagger-parser/online/) to say the spec is invalid.

mayaepps committed:
Add diffing and versioning to the RenderIndex spec

Updates RenderIndex spec with a dictionary of all node identifiers that map to the diffs (added, modified) that should be displayed against each previous version. Also adds version metadata.
